parsing/parsetree.mli

```diff
@@ -385,6 +387,10 @@ and expression_desc =
         (* [%id] *)
   | Pexp_unreachable
         (* . *)
+  | Pexp_functor of string loc * package_type option * expression
+        (* fun {M : S} -> E *)
+  | Pexp_functor_apply of expression * module_expr
+        (* E {M} *)
```

Comment:
I'm not sure I like this.
I think I'd like it better if these applications were represented with `Pexp_apply` and a dedicated `arg_label`, which is what the implicits prototype did.

Then we can "just" extend `type_application` to handle those.
In particular, with the approach you propose here, you get the following behavior:
```
# module type S = sig type t end;;
module type S = sig type t end
# let f {M:S} ~x ~y = x + y;;
val f : {M : S} -> x:int -> y:int -> int = <fun>
# f ~x:3 {Int} ~y:2;;
Error: This expression has type {M : S} -> x:int -> y:int -> int
       This is not a function; it cannot be applied.
```
whereas if you make the effort of integrating with the existing handling of applications, you get this:
```
f ~x:3 {Int} ~y:2;;
- : int = 5
```

Reply:
Note that my point was about the representation of the type syntax, which does not need to match the representation of apply.

@trefis is right that we need to share the apply constructor if we want to reorder label applications across explicits applications. This is because we need to distinguish e.g.:
```ocaml
let _ = foo ~a {X} ~b
```
from
```ocaml
let _ = (foo ~a {X}) ~b
```
In other words, function application is n-ary in OCaml even though function types are not.

@mrmr1993 is right about needing to be careful if we do allow reordering labels across explicits applications. We need to make sure that types constructors do not escape. Using `add_ints` as an example, I think that it is fine to allow:
```ocaml
let foo = add_ints ~x:1 ~y:2 {Int_add};;
```
but that this should be an error:
```ocaml
let bar = add_ints ~x:1 ~y:2;;
```

Reply:
I think the other example is more pressing, since this syntax becomes essentially an unlabelled optional argument when we introduce implicits. To simplify
```ocaml
val f : {M : S} -> a:int -> {N : S} -> ...
f ~a:1 {X} (* M = X, or M omitted and N = X? *)
```

We need to consider this now, because we don't want to make syntactically-valid programs invalid when we introduce implicits. Personally, I would prefer to err on the side of doing nothing. I can't see that being able to reorder parameters across functors is particularly useful, and it's something that can be easily added later (e.g. post-implicits) when there has been more time spent with and more thought about the syntax.

Reply:
I don't quite see how the "explicit implicit" syntax is any different from ordinary explicit arguments. You always need to match the arguments to the parameters and you do so based on position for unlabelled parameters.

Actual implicit parameters are more interesting. We need to make a decision about what forces an implicit argument to be used: either it is passing an unlabelled argument that comes after the implicit or it is passing any argument that comes after the implicit. That decision is probably affected by whether we allow labelled arguments to be reordered with explicits.

I think leaving this case out for now is fine, but I'd still be in favour of treating the applications with a single construct. OCaml applications are fundamentally n-ary and it is good to have the syntax match this.

Reply:
Using `Pexp_apply` turns out to be more complex than I expected: there is effectively no shared code between the handling of other argument kinds and modular explicits. The naive implementation seems to need
* a hack to stop reordering of labelled arguments through modular explicits
  - this ends up being quite invasive because of the need to split up arguments into groups
* a hack to disallow inferring the type of a function with explicits from the arguments
  - we can't tell whether `f {M} (Some 1)` has `f : {M : S} -> int option` or `f : {M : S} -> M.t` etc.
* conditional changes to perform a substitution instead of unification to get the result type of a modular explicit application
* some awkward code to extract the module from the argument expression
  - this includes special error handling, because typechecking alone won't check that it's a packed module expression which expands to a path
* special handling for printing, again having to extract the module path from a packed module expression.

It seems like things would work out cleaner if we can change the type of arguments to `Pexp_apply`/`Texp_apply` to e.g. `Parg_expression`/ `Parg_module`. Does this seem acceptable? Any other suggestions?

Reply:
As in:
```
type argument =
  | Parg_expression of arg_label * expression
  | Parg_module of Path.t
...
type expression =
...
  | Pexp_apply of expression * argument list 
...
```
That looks good to me. I think that is what I was actually picturing in the above discussion. The important bit was being able to treat the whole application as a single AST node, not keeping the old structure. You could maybe make `arg_label` a GADT indexed by the sort of argument that is needed, but that probably just adds complexity for no gain.

Reply:
Disclaimer: what follows is not prescriptive, but rather a FYI description of the choices I made in my rebase/rewrite of the implicits prototype:

- I don't quite have the AST Leo describes, but I think that's because he wrote `P` (for parsetree) where he should have written `T`, here is what I do have (sorry for the length, but hopefully that should be enough to disambiguate):
  ```ocaml
  (* parsing *)
  type Asttypes.arg_label = Nolabel | Labelled of string | Optional of string (* unchanged *)
  type Asttypes.apply_label =
    | Papp_nolabel
    | Papp_labelled of string
    | Papp_optional of string
    | Papp_implicit
  
  type Parsetree.expression_desc =
    | Pexp_fun of arg_label * expression option * pattern * expression (* unchanged *)
    | Pexp_functor of string * package_type * expression
    | Pexp_apply of expression * (apply_label * expression) list
    | ...

  (* typing *)
  type Types.type_desc =
    | Tarrow of arg_label * type_expr * type_expr * commutable (* unchanged *)
    | Tfunctor of Ident.t * type_expr (* perhaps this should be a package *) * type_expr * commutable
    | ...

  type Typedtree.argument =
    | Normal of arg_label * expression option
    | Implicit of expression option ref (* Mutable to defer implicit instance selection *)

  type Typedtree.expression_desc =
    | Texp_function of { arg_label : arg_label; param : Ident.t; cases : value case list; partial : partial; } (* unchanged *)
    | Texp_functor of Ident.t * package_type * expression
    | Texp_apply of expression * argument list
  ```
  Clearly `Typedtree.argument` needs some love: Leo's constructor names are better, and I'm not sure whether we want to restrict the module argument to a path here, but the general shape of things is the same.
- > a hack to disallow inferring the type of a function with explicits from the arguments
  
  I don't think you should ever infer the type of a function taking an explicit as argument. So if we take your example: if you start with `f : 'a`, you shouldn't allow passing `{M}` as argument. But if you start with `f : {M : S} -> 'a` I'm not sure there is a problem? Both `M.t` and `int option` should be equal under the `Tfunctor` (at least that's what I'm guessing from your message), and if it turns out that the `'a` is at an outer level and `M.t` can't be chosen because it would escape, then I believe at that point you should still know that they are equal and you can simply expand to `int option`.
  Does that make sense?
- > a hack to stop reordering of labelled arguments through modular explicits
I think I'd rather avoid that if we can. Is the reason to forbid it that it's the only way to prevent `let bar = add_ints ~x:1 ~y:2;;` (discussed above) from typechecking?

I don't have precise things to say about the rest, that code (`type_application`) is indeed rather involved an adding explicits to the mix is not exactly making anything clearer.

Reply:
@trefis If I'm reading your code right, it looks like it is still using a `expression` to represent explicit module arguments. I think that is probably a mistake since presumably it requires encoding the module paths as e.g. FCMs (since we've gone to the trouble of minting this new acronym I guess I can now use it in all future discussions).

Reply:
Updated. Functor application now uses `Pexp_apply` via a `Parg_module` constructor, as above.